src/aims_ui/api_interaction.py

```diff
@@ -3,43 +3,137 @@
+    url = app.config.get('API_URL') + url + all_user_input.get(called_from, '')
     params = get_params(all_user_input)
-    r = requests.get(url, params=params, headers=header, )
+    r = requests.get(
+        url,
+        params=params,
+        headers=header,
+    )
 
   elif called_from == 'singlesearch':
     url = app.config.get('API_URL') + url
     proposed_params = {k: v for k, v in all_user_input.items() if v}
     params = get_params(all_user_input)
-    r = requests.get(url, params=params, headers=header, )
+    r = requests.get(
+        url,
+        params=params,
+        headers=header,
+    )
 
   return r
+
+
+def multiple_address_match(file, all_user_input, app, download=False):
+  final_csv = 'id, inputAddress, matchedAddress, uprn, matchType, confidenceScore, documentScore, rank\n\r'
+  ths = [{'value': x, 'ariaSort': None} for x in final_csv.split(',')]
+  trs = []
+
+  contents = file.readlines()
+  if download:
+    proxy = StringIO()
+    writer = csv.writer(proxy)
+    writer.writerow(final_csv.split(','))
+
+  line_count = 0
+  for line in contents:
+    line = line.strip().decode("utf-8")
+    given_id = line.split(',')[0]
+    address_to_lookup = line.split(',')[1]
+    all_user_input['input'] = address_to_lookup
+
+    result = api(
+        '/addresses',
+        'singlesearch',
+        all_user_input,
+    )
+
+    matched_addresses = get_addresses(result.json(), 'singlesearch', app)
+    match_type = '<p style="background-color:orange;">M</p>' if len(
+        matched_addresses
+    ) > 1 else '<p style="background-color:Aquamarine;">S</p>'
+    rank = 1
+    for adrs in matched_addresses:
+      if download:
+        writer.writerow([
+            given_id, address_to_lookup, adrs.formatted_address_nag.value,
+            adrs.uprn.value, match_type, adrs.confidence_score.value,
+            adrs.underlying_score.value, rank
+        ])
+        line_count = line_count + 1
+      else:
+        trs.append({
+            'tds': [
+                {
+                    'value': given_id
+                },
+                {
+                    'value': address_to_lookup
+                },
+                {
+                    'value': adrs.formatted_address_nag.value
+                },
+                {
+                    'value': adrs.uprn.value
+                },
+                {
+                    'value': match_type
+                },
+                {
+                    'value': adrs.confidence_score.value
+                },
+                {
+                    'value': adrs.underlying_score.value
+                },
+                {
+                    'value': rank
+                },
+            ]
+        })
+        final_csv=final_csv+ f'{given_id},{address_to_lookup},{adrs.formatted_address_nag.value},' +\
+            f'{adrs.uprn.value},{match_type},{adrs.confidence_score.value},{adrs.underlying_score.value},{rank}\n\r'
+      rank = rank + 1
+
+  if download:
```

Comment:
Likewise, this ending part could be abstracted away just by defining a `def finalize` for each method up above, and then calling it here.

Reply:
I have included this in the latest push too :)